### PR TITLE
Llama: fix custom 4D masks

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

Fixes the issue raised by @poedator in [this comment](https://github.com/huggingface/transformers/pull/29753#issuecomment-2014530814).

Causal mask is now of shape `[..., seq_len, full_len]`, as opposed to `[..., full_len, full_len]`. This means custom 4D attention masks are now the whole causal mask, so we don't need a sliced copy -- we can copy the whole thing :) 

This PR also expands the support of custom 4D attention mask: we can pass both the full mask (`[..., full_len, full_len]`) or the partial mask (`[..., seq_len, full_len]`).

### CODE REVIEW

File: src/transformers/models/cohere/modeling_cohere.py [hunk 1]
```diff
@@ -963,36 +963,34 @@ def _update_causal_mask(self, attention_mask, input_tensor, cache_position):
...
 
-        causal_mask = torch.full((sequence_length, target_length), fill_value=min_dtype, dtype=dtype, device=device)
-        if sequence_length != 1:
-            causal_mask = torch.triu(causal_mask, diagonal=1)
-        causal_mask *= torch.arange(target_length, device=device) > cache_position.reshape(-1, 1)
-        causal_mask = causal_mask[None, None, :, :].expand(input_tensor.shape[0], 1, -1, -1)
-        if attention_mask is not None:
-            causal_mask = causal_mask.clone()  # copy to contiguous memory for in-place edit
-            if attention_mask.dim() == 2:
+        if attention_mask is not None and attention_mask.dim() == 4:
```
Comment (gante) [inline src/transformers/models/cohere/modeling_cohere.py:967]: reordered the logic: custom 4D masks are now a superset of the default mask, so we don't need to create the default mask first :)

File: src/transformers/models/cohere/modeling_cohere.py [hunk 2]
```diff
@@ -963,36 +963,34 @@ def _update_causal_mask(self, attention_mask, input_tensor, cache_position):
...
+            # we can pass both the full 4D mask (i.e. [..., full_len, full_len]) and a 4D mask with the same shape
+            # as the causal mask (i.e. [..., seq_len, full_len])
+            mask_slice = (attention_mask.eq(0.0)).to(dtype=dtype) * min_dtype
+            if attention_mask.shape[-2] == cache_position[0] + sequence_length:
+                offset = cache_position[0]
+                mask_slice = mask_slice[..., offset : offset + sequence_length, :]
+            causal_mask = mask_slice
+        else:
```
Comment (gante) [inline src/transformers/models/cohere/modeling_cohere.py:975]: This `else` has no changes. Only the `if attention_mask is not None and attention_mask.dim() == 4:` is different.

File: tests/test_modeling_utils.py [hunk 3]
```diff
@@ -2208,17 +2208,28 @@ def test_partial_stacked_causal_mask(self):
...
         )
-
         decoded_1b = [
             self.tokenizer.decode(t)
             for t in outs_1b.logits.argmax(-1)[0, torch.where(position_ids_1 == position_ids_1.max())[1] - part_a]
         ]
-
         self.assertEqual(decoded_0, decoded_1b)
+
+        # Case 2: we pass a 4D attention mask regarding the full sequence length (i.e. [..., full_len, full_len])
```
Comment (gante) [inline tests/test_modeling_utils.py:2224]: Added this test case (we can now pass full custom 4D attention masks)

File: src/transformers/models/cohere/modeling_cohere.py [hunk 4]
```diff
@@ -963,36 +963,34 @@ def _update_causal_mask(self, attention_mask, input_tensor, cache_position):
...
-        causal_mask = causal_mask[None, None, :, :].expand(input_tensor.shape[0], 1, -1, -1)
-        if attention_mask is not None:
-            causal_mask = causal_mask.clone()  # copy to contiguous memory for in-place edit
-            if attention_mask.dim() == 2:
+        if attention_mask is not None and attention_mask.dim() == 4:
+            # we can pass both the full 4D mask (i.e. [..., full_len, full_len]) and a 4D mask with the same shape
+            # as the causal mask (i.e. [..., seq_len, full_len])
+            mask_slice = (attention_mask.eq(0.0)).to(dtype=dtype) * min_dtype
+            if attention_mask.shape[-2] == cache_position[0] + sequence_length:
+                offset = cache_position[0]
```
Comment (ArthurZucker) [inline src/transformers/models/cohere/modeling_cohere.py:972]:
```suggestion
            offset = cache_position[0]
            if attention_mask.shape[-2] == offset + sequence_length:
```

File: tests/test_modeling_utils.py [hunk 5]
```diff
@@ -2208,17 +2208,28 @@ def test_partial_stacked_causal_mask(self):
...
+        position_ids_1c = position_ids_1[:, part_a:]
+        mask_1c = mask_1
+        outs_1c = self.model.forward(
+            input_1c, attention_mask=mask_1c.bool(), position_ids=position_ids_1c, past_key_values=past_key_values_a
+        )
+        decoded_1c = [
+            self.tokenizer.decode(t)
+            for t in outs_1c.logits.argmax(-1)[0, torch.where(position_ids_1 == position_ids_1.max())[1] - part_a]
+        ]
+        self.assertEqual(decoded_0, decoded_1c)
```
Comment (ArthurZucker) [inline tests/test_modeling_utils.py:2235]: Can we make sur this is triggered when running LlamaTests or Gemma etc ?

**Review by ArthurZucker — COMMENTED** [review]
Thanks LGTM just want to always trigger the tests

**Comment by poedator** [comment]
Thanks @gante ! 
I tested the code and it works.
My only suggestion is to cover the 4D masks in the [documentation for Llama](https://huggingface.co/docs/transformers/main/en/model_doc/llama) and other models.

**Comment by gante** [comment]
> My only suggestion is to cover the 4D masks in the [documentation for Llama](https://huggingface.co/docs/transformers/main/en/model_doc/llama) and other models.

@poedator would you like to open a PR with that? As a user, you'll probably have cool examples in mind!

**Comment by poedator** [comment]
> > My only suggestion is to cover the 4D masks in the [documentation for Llama](https://huggingface.co/docs/transformers/main/en/model_doc/llama) and other models.
> 
> @poedator would you like to open a PR with that? As a user, you'll probably have cool examples in mind!

will try, but not this week...

File: tests/models/llama/test_modeling_llama.py [hunk 6]
```diff
@@ -735,3 +736,138 @@ def test_model_7b_logits(self):
         ]
         infilling = tokenizer.batch_decode(generated_ids)
         self.assertEqual(infilling, EXPECTED_INFILLING)
+
+
+@slow
```
Comment (gante) [inline tests/models/llama/test_modeling_llama.py:741]: This set of slow tests was now moved to the llama test file -> if we run the slow llama tests, this will now be triggered

File: tests/test_modeling_common.py [hunk 7]
```diff
@@ -4027,6 +4027,101 @@ def test_flash_attn_2_from_config(self):
 
                 self.assertFalse(fa2_correctly_converted)
 
+    def _get_custom_4d_mask_test_data(self):
```
Comment (gante) [inline tests/test_modeling_common.py:4030]: This set of tests are now:
1. part of the mixing, so they are run on all push commits
2. a fast test, using `model = model_class(config)` from the test config
3. triggered by `model_class._supports_cache_class == True` -- recent LLMs [llama, cohere, gemma, mistral, mixtral, starcoder2, ...] have this attribute set to `True` and are 4D mask-compatible. Older models are often not compatible. Over time, as we spread the cache refactor, this test will be run on those classes as well 👀

**Comment by gante** [comment]
@ArthurZucker ready for a re-review (test rework) -- we now have on push tests for all recent models + custom 4D mask :)

**Comment by poedator** [comment]
@gante ,
May I suggest adding extra tests for 4D masks with StaticCache? I am concerned that StaticCache code may handle the custom masks differently.
Here are the additional methods for `Mask4DTestHard`:
https://gist.github.com/poedator/f1c15551d202df2682c65c1bbdcb1c07

I made the cache longer than the masks and padded the masks to the cache length. is this the correct way?

**Comment by ArthurZucker** [comment]
Sorry for the delay, let's rebase on main as well

**Review by ArthurZucker — APPROVED** [review]
Very good! Let's rebase on main, #30047  was merged, and run slow tests!

**Comment by poedator** [comment]
please, please merge this PR - I need it for my speculative decoding paper project! The 4D masks are essential for it. 
@gante @ArthurZucker

**Comment by ArthurZucker** [comment]
Sorry just got back to github 😓 could you rebase!

**Comment by poedator** [comment]
> Sorry just got back to github 😓 could you rebase!
Too bad it did not make it into 4.40 (

I rebased this PR in new one #30348 
Apparently there is a conflict with `AttentionMaskConverter._ignore_causal_mask_sdpa()` Please see the new PR with details. Need your guidance to fix this.

**Comment by gante** [comment]
(a few tests broke on the rebase, need to revisit the changes here 👀 )

**Comment by gante** [comment]
Closing in favor of #30348
